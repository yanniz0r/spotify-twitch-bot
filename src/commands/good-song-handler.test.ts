import SpotifyWebApi from "spotify-web-api-node";
import TwitchBot, { TwitchUser } from "../twitch-bot";
import TwitchConnectionAdapter from "../twitch-connection-adapter";
import GoodSongHandler from "./good-song-handler";

jest.mock("spotify-web-api-node");

const whatDoesTheFoxSay = {
  id: "tfwdtfs",
  uri: "tfwdtfs",
  name: "The Fox (What Does the Fox Say?)",
  artists: [
    {
      name: "Ylvis",
    },
  ],
}

describe('GoodSongHandler', () => {

  let handler: GoodSongHandler;
  let spotify: SpotifyWebApi;
  let bot: TwitchBot;
  let adapter: TwitchConnectionAdapter;
  let user: TwitchUser = {
    username: "yanniz0r",
  };

  beforeEach(() => {
    adapter = {
      register: jest.fn(),
      sendMessage: jest.fn(),
    };
    bot = new TwitchBot(adapter);
    spotify = new SpotifyWebApi();
    handler = new GoodSongHandler('test-playlist-id', spotify);
  });

  it ('sends a corresponding message when no track is running', async () => {
    spotify.getMyCurrentPlayingTrack = jest.fn().mockReturnValue({
      body: {
        item: null
      },
    });
    spotify.getPlaylistTracks = jest.fn().mockReturnValue({
      body: {
        items: []
      },
    });
    await handler.handle(bot, {
      channel: '#test',
      text: '!gutersong',
      user,
    })
    expect(spotify.addTracksToPlaylist).toMatchSnapshot();
  })

  it ('adds a new song to the playlist and sends corresponding message', async () => {
    spotify.getMyCurrentPlayingTrack = jest.fn().mockReturnValue({
      body: {
        item: whatDoesTheFoxSay
      },
    });
    spotify.getPlaylistTracks = jest.fn().mockReturnValue({
      body: {
        items: []
      },
    });
    spotify.addTracksToPlaylist = jest.fn();
    await handler.handle(bot, {
      channel: '#test',
      text: '!gutersong',
      user,
    })
    expect(spotify.addTracksToPlaylist).toMatchSnapshot();
  })

  it ('does not add duplicates and sends corresponding message', async () => {
    spotify.getMyCurrentPlayingTrack = jest.fn().mockReturnValue({
      body: {
        item: whatDoesTheFoxSay
      },
    });
    spotify.getPlaylistTracks = jest.fn().mockReturnValue({
      body: {
        items: [{
          track: whatDoesTheFoxSay
        }]
      },
    });
    spotify.addTracksToPlaylist = jest.fn();
    await handler.handle(bot, {
      channel: '#test',
      text: '!gutersong',
      user,
    })
    expect(spotify.addTracksToPlaylist).not.toBeCalled();
    expect(spotify.addTracksToPlaylist).toMatchSnapshot();
  })

})