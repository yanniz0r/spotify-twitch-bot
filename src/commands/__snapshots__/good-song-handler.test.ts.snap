// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoodSongHandler adds a new song to the playlist and sends corresponding message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "test-playlist-id",
      Array [
        "tfwdtfs",
      ],
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`GoodSongHandler does not add duplicates and sends corresponding message 1`] = `[MockFunction]`;

exports[`GoodSongHandler sends a corresponding message when no track is running 1`] = `[MockFunction]`;
